lib/AST/ASTImporter.cpp

```diff
@@ -8154,15 +8154,22 @@ Expected<FileID> ASTImporter::Import(FileID FromID) {
       // than mmap the files several times.
       const FileEntry *Entry =
           ToFileManager.getFile(Cache->OrigEntry->getName());
-      if (!Entry)
-        // FIXME: Use a new error kind?
-        return llvm::make_error<ImportError>(ImportError::Unknown);
-      ToID = ToSM.createFileID(Entry, *ToIncludeLoc,
-                               FromSLoc.getFile().getFileCharacteristic());
-    } else {
+      // FIXME: The filename may be a virtual name that does probably not
+      // point to a valid file and we get no Entry here. In this case try with
+      // the memory buffer below.
+      if (Entry)
+        ToID = ToSM.createFileID(Entry, *ToIncludeLoc,
+                                 FromSLoc.getFile().getFileCharacteristic());
+    }
+    if (ToID.isInvalid()) {
       // FIXME: We want to re-use the existing MemoryBuffer!
+      bool Invalid = true;
       const llvm::MemoryBuffer *FromBuf =
-          Cache->getBuffer(FromContext.getDiagnostics(), FromSM);
+          Cache->getBuffer(FromContext.getDiagnostics(), FromSM, SourceLocation{}, &Invalid);
+      if (!FromBuf || Invalid)
+        // FIXME: Use a new error kind?
+        return llvm::make_error<ImportError>(ImportError::Unknown);
```

Comment:
I think we should use a new error kind indeed, because in `CrossTranslationUnit.cpp` we have an `llvm_unreachable` if the error is `ImportError::Unknown`.
```
                      case ImportError::Unknown:
                        llvm_unreachable("Unknown import error happened.");
                        break;
```

Reply:
If we encounter this error a new error kind can be introduced, but for now we can assume that this case does not occur at all (if yes the assert above will fail).

Reply:
okay.